### ISSUE TITLE
To control interval or storage of stats.ucl

### ISSUE THREAD

**Issue by satanTime** [rspamd/rspamd#2680]
### Classification (Please choose *one* option):

* [ ] Crash/Hang/Data loss
* [x] WebUI/Usability
* [ ] Serious bug
* [ ] Ordinary bug
* [ ] Feature
* [ ] Enhancement

### Reproducibility (Please choose *one* option):

* [X] Always
* [ ] Sometimes
* [ ] Rarely
* [ ] Unable
* [ ] I didn’t try
* [ ] Not applicable

### Rspamd version:
1.8.3

### Operation system, CPU:
Dockerized Debian Stretch, 1CPU

### Description (Please provide a descriptive summary of the issue):
In my case it's very simple. Once I got some emails processed the file isn't updated.
When I restart container I get zeros for everything.
Time to time the file is updated, but in very long intervals, and I can get very outdated stats after reboot.

### Compile errors (if any):
not really

### Expected results:
should be updated /var/lib/rspamd/stats.ucl with fresh counters

### Actual results:
 /var/lib/rspamd/stats.ucl contains {"scanned":0,"learned":0,"connections":0,"control_connections":0}

### Configuration (e.g. `rspamadm configdump module`):
*** Section worker ***
controller {
    password = "...";
    secure_ip = "worker-controller.map";
    stats_path = "/var/lib/rspamd/stats.ucl";
    count = 1;
    bind_socket = "*:11334";
}
*** End of section worker ***

### Additional information:
When I make file empty with zero size, after I resize an email it writes there 
{"scanned":0,"learned":0,"connections":0,"control_connections":0}
when it should be
{"scanned":1,"learned":0,"connections":0,"control_connections":0}
I guess.

Is there way to move stats to redis for example? Or to add something to logging to see more detailed information? Thanks.

**Comment by vstakhov** [comment]
It is saved once per minute.

**Comment by satanTime** [comment]
In my case it saves always {"scanned":0,"learned":0,"connections":0,"control_connections":0}, when UI shows real numbers.
When I truncate file to zero size then after some time it has again {"scanned":0,"learned":0,"connections":0,"control_connections":0}

that's the problem. Unfortunately I wasn't able to find flags for logging module how to enable verbose output about update of the file.

**Comment by satanTime** [comment]
Ah, control_connections changes between saves.

**Comment by satanTime** [comment]
Hi @vstakhov, after 24 hours I have next stats in UI:

188
Scanned

73
Clean

11
Greylist

6
Probable

98
Reject

116
Learned

and stats.ucl has next stats:
root@office:/var/lib/rspamd# ls -lah stats.ucl
-rw-r--r-- 1 _rspamd _rspamd 65 Dec 13 08:06 stats.ucl
root@office:/var/lib/rspamd# cat stats.ucl
{"scanned":0,"learned":0,"connections":0,"control_connections":1}r

as you can see file wasn't updated since yesterday.

Any way I could debug why it is so?
Thanks.

**Comment by satanTime** [comment]
Could it be caused that `/var/lib/rspamd` is mounted?

**Comment by moisseev** [comment]
> It is saved once per minute.

It is not true. I guess it saved only on rspamd stop.

**Comment by satanTime** [comment]
Aha, good to know. I'll test it outside of docker to double check reality.

**Comment by satanTime** [comment]
On my side the issue was cased due to wrong termination process when main script didn't send term signal correctly to rspamd for graceful shutdown and stats weren't written.
I've fixed that part and now the file is written on exit.

**Comment by feddar** [comment]
@satanTime how did you fixed that? Thanks
> On my side the issue was cased due to wrong termination process when main script didn't send term signal correctly to rspamd for graceful shutdown and stats weren't written.
> I've fixed that part and now the file is written on exit.

**Comment by satanTime** [comment]
Hi @feddar, I use supervisord inside and added rule to send stopsignal=TERM and stopwaitsecs=30. In docker I added stop_grace_period: 1m.

Also I use docker-entrypoint pattern and added next code to forward termination:

```bash
#!/bin/bash
echo "${0}:start"

child_pid=0
parent_pid=$$

catch_exits() {
	echo "${0}:stop ${child_pid}"
    kill ${child_pid} &
    wait
    if [[ -f /var/run/supervisor.pid ]]; then
        child_pid=$(cat /var/run/supervisor.pid);
    	echo "${0}:stop ${child_pid}"
        kill ${child_pid} &
        wait
        tail --pid=${child_pid} -f /dev/null 2> /dev/null > /dev/null
    fi
	echo "${0}:exit"
	exit 0
}
trap catch_exits SIGTERM

set -e
command=${@}

for file in /docker-entrypoint.d/*
do
  sh -c "${file}"
done

fork() {
    echo "${command}"
    sh -c "${command}"
}

fork &
child_pid=$!
echo "${0}:child ${child_pid}"
wait ${child_pid}
```

**Comment by feddar** [comment]
Wow @satanTime seems to be complicated!  I didn't imagine...Thank you very much for sharing your code. I'm using Alpine as base image, i will try to get it working using [tini](https://github.com/krallin/tini)

**Comment by satanTime** [comment]
Yeah, my problem was that docker-entrypoint.sh didn't forward term signals to terminate other processes correctly. In your case if you use alphine and run rspam there directly without middle scripts - it should work correctly out of the box, only stop_grace_period: 1m could be used to be sure rspamd had enough time to finish work gracefully.

**Comment by feddar** [comment]
Unfortunately i have to use an intermediate script because i have to check (and wait) that other containers (like clamav, dcc, pyzor, php-fpm, etc) are up, before fire up rspamd. But your work can give me some ideas to work on.
